chalk-ir/src/lib.rs

```diff
@@ -1175,17 +1176,60 @@ pub enum Constraint<TF: TypeFamily> {
 }
 
 /// A mapping of inference variables to instantiations thereof.
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Fold, Hash, HasTypeFamily)]
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, HasTypeFamily)]
 pub struct Substitution<TF: TypeFamily> {
     /// Map free variable with given index to the value with the same
     /// index. Naturally, the kind of the variable must agree with
     /// the kind of the value.
-    pub parameters: Vec<Parameter<TF>>,
+    parameters: TF::InternedSubstitution,
```

Comment:
At this point, why not just replace `Substitution` with an associated type?

Reply:
This pattern is what we use for all the "interned things". Having the newtype'd wrapper allows us to implement traits like `Fold` and add methods. You can't *quite* make `Fold` work without the newtype'd wrapper. The other advantage is consistency: you always write `Foo<TF>`, never `TF::Foo`, in regular code.